### PR TITLE
V nikija/app complete sample v4

### PR DESCRIPTION
**Pull request by AbhijeetThakur-msft**
Changes for app migration from bot framework v3 to v4

### CODE REVIEW

File: samples/app-complete-sample/csharp/template-bot-master-csharp/src/dialogs/examples/moderate/PromptDialog.cs [hunk 1]
```diff
@@ -13,87 +17,97 @@ namespace Microsoft.Teams.TemplateBotCSharp.Dialogs
...
+        private static List<Choice> options = new List<Choice>()
+            {
+                new Choice(Strings.PlayGameChoice1) { Synonyms = new List<string> { Strings.PlayGameChoice1 } },
+                new Choice(Strings.PlayGameChoice2) { Synonyms = new List<string> { Strings.PlayGameChoice2 } },
+                new Choice(Strings.PlayGameWrongChoice) { Synonyms = new List<string> { Strings.PlayGameWrongChoice } }
+            };
+        private static List<Choice> choiceOptions = new List<Choice>()
+            {
+                new Choice(Strings.OptionYes) { Synonyms = new List<string> { Strings.OptionYes } },
+                new Choice(Strings.OptionNo) { Synonyms = new List<string> { Strings.OptionNo } },
```
Comment (v-chetsh) [inline samples/app-complete-sample/csharp/template-bot-master-csharp/src/dialogs/examples/moderate/PromptDialog.cs:32]: Alignment is not proper

File: samples/app-complete-sample/csharp/template-bot-master-csharp/src/dialogs/examples/teams/AtMentionDialog.cs [hunk 2]
```diff
@@ -12,25 +14,39 @@ namespace Microsoft.Teams.TemplateBotCSharp.Dialogs
...
+            InitialDialogId = nameof(WaterfallDialog);
+            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[]
+            {
+                BeginAtMentionDialogAsync,
+            }));
+        }
+
+        private async Task<DialogTurnResult> BeginAtMentionDialogAsync(
+WaterfallStepContext stepContext,
+CancellationToken cancellationToken = default(CancellationToken))
```
Comment (v-chetsh) [inline samples/app-complete-sample/csharp/template-bot-master-csharp/src/dialogs/examples/teams/AtMentionDialog.cs:32]: Alignment is not proper

File: samples/app-complete-sample/csharp/template-bot-master-csharp/src/dialogs/examples/teams/DeepLinkStaticTabDialog.cs [hunk 3]
```diff
@@ -24,33 +25,46 @@ public class DeepLinkStaticTabDialog : IDialog<object>
...
             {
-                throw new ArgumentNullException(nameof(context));
-            }
+                BeginDeepLinkStaticTabDialogAsync,
+            }));
+        }
 
+        private async Task<DialogTurnResult> BeginDeepLinkStaticTabDialogAsync(
+WaterfallStepContext stepContext,
+CancellationToken cancellationToken = default(CancellationToken))
```
Comment (v-chetsh) [inline samples/app-complete-sample/csharp/template-bot-master-csharp/src/dialogs/examples/teams/DeepLinkStaticTabDialog.cs:41]: Alignment is not proper

File: samples/app-complete-sample/csharp/template-bot-master-csharp/src/dialogs/examples/teams/DisplayCardsDialog.cs [hunk 4]
```diff
@@ -1,92 +1,132 @@
...
+        protected readonly IStatePropertyAccessor<RootDialogState> _conversationState;
+        private static List<Choice> options = new List<Choice>()
+            {
+                new Choice(Strings.DisplayCardHeroCard) { Synonyms = new List<string> { Strings.DisplayCardHeroCard } },
+                new Choice(Strings.DisplayCardThumbnailCard) { Synonyms = new List<string> { Strings.DisplayCardThumbnailCard } },
+                new Choice(Strings.DisplayCardO365ConnectorCardDefault) { Synonyms = new List<string> { Strings.DisplayCardO365ConnectorCardDefault } },
+                new Choice(Strings.DisplayCardO365ConnectorCard2) { Synonyms = new List<string> { Strings.DisplayCardO365ConnectorCard2 } },
+                new Choice(Strings.DisplayCardO365ConnectorCard3) { Synonyms = new List<string> { Strings.DisplayCardO365ConnectorCard3 } },
+                new Choice(Strings.DisplayCardO365ConnectorActionableCardDefault) { Synonyms = new List<string> { Strings.DisplayCardO365ConnectorActionableCardDefault } },
+                new Choice(Strings.DisplayCardO365ConnectorActionableCard2) { Synonyms = new List<string> { Strings.DisplayCardO365ConnectorActionableCard2 } },
```
Comment (v-chetsh) [inline samples/app-complete-sample/csharp/template-bot-master-csharp/src/dialogs/examples/teams/DisplayCardsDialog.cs:26]: Alignment is not proper

File: samples/app-complete-sample/csharp/template-bot-master-csharp/src/dialogs/examples/teams/FetchRosterDialog.cs [hunk 5]
```diff
@@ -11,31 +16,41 @@ namespace Microsoft.Teams.TemplateBotCSharp.Dialogs
...
+            InitialDialogId = nameof(WaterfallDialog);
+            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[]
+            {
+                BeginFetchRosterDialogAsync
+            }));
+        }
+
+            private async Task<DialogTurnResult> BeginFetchRosterDialogAsync(
+    WaterfallStepContext stepContext,
+    CancellationToken cancellationToken = default(CancellationToken))
```
Comment (v-chetsh) [inline samples/app-complete-sample/csharp/template-bot-master-csharp/src/dialogs/examples/teams/FetchRosterDialog.cs:35]: Alignment is not proper

File: samples/app-complete-sample/csharp/template-bot-master-csharp/src/dialogs/examples/teams/FetchTeamsInfoDialog.cs [hunk 6]
```diff
@@ -13,41 +17,54 @@ namespace Microsoft.Teams.TemplateBotCSharp.Dialogs
...
-                throw new ArgumentNullException(nameof(context));
-            }
+                BeginFetchTeamsInfoDialogAsync,
+            }));
+        }
 
-            var team = context.Activity.GetChannelData<TeamsChannelData>().Team;
+        private async Task<DialogTurnResult> BeginFetchTeamsInfoDialogAsync(
+WaterfallStepContext stepContext,
+CancellationToken cancellationToken = default(CancellationToken))
```
Comment (v-chetsh) [inline samples/app-complete-sample/csharp/template-bot-master-csharp/src/dialogs/examples/teams/FetchTeamsInfoDialog.cs:35]: Alignment is not proper

File: samples/app-complete-sample/csharp/template-bot-master-csharp/src/dialogs/examples/teams/HelpDialog.cs [hunk 7]
```diff
@@ -11,20 +14,34 @@ namespace Microsoft.Teams.TemplateBotCSharp.Dialogs
...
+            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[]
             {
-                throw new ArgumentNullException(nameof(context));
+                BeginHelpDialogAsync,
+            }));
+        }
+
+        private async Task<DialogTurnResult> BeginHelpDialogAsync(
+WaterfallStepContext stepContext,
+CancellationToken cancellationToken = default(CancellationToken))
```
Comment (v-chetsh) [inline samples/app-complete-sample/csharp/template-bot-master-csharp/src/dialogs/examples/teams/HelpDialog.cs:32]: Alignment is not proper

File: samples/app-complete-sample/csharp/template-bot-master-csharp/src/dialogs/examples/teams/ProactiveMsgTo1to1Dialog.cs [hunk 8]
```diff
@@ -1,34 +1,54 @@
...
             {
-                throw new ArgumentNullException(nameof(context));
+                BeginProactiveMsgTo1to1DialogAsync,
+            }));
+        }
+
+        private async Task<DialogTurnResult> BeginProactiveMsgTo1to1DialogAsync(
+WaterfallStepContext stepContext,
+CancellationToken cancellationToken = default(CancellationToken))
+        {
```
Comment (v-chetsh) [inline samples/app-complete-sample/csharp/template-bot-master-csharp/src/dialogs/examples/teams/ProactiveMsgTo1to1Dialog.cs:34]: Alignment is not proper

File: samples/app-complete-sample/csharp/template-bot-master-csharp/src/dialogs/examples/teams/UpdateCardMsgSetupDialog.cs [hunk 9]
```diff
@@ -12,32 +15,46 @@ namespace Microsoft.Teams.TemplateBotCSharp.Dialogs
...
+            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[]
             {
-                throw new ArgumentNullException(nameof(context));
+                BeginUpdateCardMsgSetupDialogAsync,
+            }));
+        }
+
+        private async Task<DialogTurnResult> BeginUpdateCardMsgSetupDialogAsync(
+WaterfallStepContext stepContext,
+CancellationToken cancellationToken = default(CancellationToken))
```
Comment (v-chetsh) [inline samples/app-complete-sample/csharp/template-bot-master-csharp/src/dialogs/examples/teams/UpdateCardMsgSetupDialog.cs:34]: Alignment is not proper

File: samples/app-complete-sample/csharp/template-bot-master-csharp/src/dialogs/examples/teams/UpdateTextMsgDialog.cs [hunk 10]
```diff
@@ -1,45 +1,65 @@
...
+            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[]
             {
-                throw new ArgumentNullException(nameof(context));
+                BeginUpdateTextMsgDialogAsync,
+            }));
+        }
+
+        private async Task<DialogTurnResult> BeginUpdateTextMsgDialogAsync(
+WaterfallStepContext stepContext,
+CancellationToken cancellationToken = default(CancellationToken))
```
Comment (v-chetsh) [inline samples/app-complete-sample/csharp/template-bot-master-csharp/src/dialogs/examples/teams/UpdateTextMsgDialog.cs:32]: Alignment is not proper

File: samples/app-complete-sample/csharp/template-bot-master-csharp/src/dialogs/examples/teams/UpdateTextMsgSetupDialog.cs [hunk 11]
```diff
@@ -1,37 +1,53 @@
...
-                throw new ArgumentNullException(nameof(context));
+                BeginUpdateTextMsgSetupDialogAsync,
+            }));
+        }
+
+        private async Task<DialogTurnResult> BeginUpdateTextMsgSetupDialogAsync(
+WaterfallStepContext stepContext,
+CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (stepContext == null)
```
Comment (v-chetsh) [inline samples/app-complete-sample/csharp/template-bot-master-csharp/src/dialogs/examples/teams/UpdateTextMsgSetupDialog.cs:34]: Alignment is not proper

File: samples/app-complete-sample/csharp/template-bot-master-csharp/utility/TemplateUtility.cs [hunk 12]
```diff
@@ -43,16 +42,35 @@ public static string GetLocale(Activity activity)
...
 
-        public static ComposeExtensionAttachment CreateComposeExtensionCardsAttachmentsSelectedItem(WikiHelperSearchResult wikiResult, string selectedType)
+        public static async Task SaveBotUserDataObject(IStatePropertyAccessor<UserData> userState, ITurnContext<IInvokeActivity> turnContext, List<WikiHelperSearchResult> historySearchWikiResult)
         {
-            return GetComposeExtensionMainResultAttachment(wikiResult, selectedType).ToComposeExtensionAttachment();
+            var currentState = await userState.GetAsync(turnContext, () => new UserData());
+            currentState.ComposeExtensionSelectedResults = historySearchWikiResult;
+            await userState.SetAsync(turnContext, currentState);
+        }
+        public static MessagingExtensionAttachment CreateComposeExtensionCardsAttachments(WikiHelperSearchResult wikiResult, string selectedType)
```
Comment (v-chetsh) [inline samples/app-complete-sample/csharp/template-bot-master-csharp/utility/TemplateUtility.cs:65]: Add empty line before it.

File: samples/app-complete-sample/csharp/template-bot-master-csharp/utility/WikipediaComposeExtension.cs [hunk 13]
```diff
@@ -22,42 +22,74 @@ public class WikipediaComposeExtension
...
-        const string ComposeExtensionSelectedResultsKey = "ComposeExtensionSelectedResults";
         const string MaxComposeExtensionHistoryCountKey = "MaxComposeExtensionHistoryCount";
         public static HttpClient Client = new HttpClient();
+        protected readonly IStatePropertyAccessor<UserData> _userState;
+        public WikipediaComposeExtension(IStatePropertyAccessor<UserData> userState)
+        {
+            _userState = userState;
+        }
+        public async Task<ImageResult> SearchWikiImage(Search wikiSearch)
+        {
```
Comment (v-chetsh) [inline samples/app-complete-sample/csharp/template-bot-master-csharp/utility/WikipediaComposeExtension.cs:33]: Empty line before it. Also add summary.

File: samples/app-complete-sample/csharp/template-bot-master-csharp/utility/WikipediaComposeExtension.cs [hunk 14]
```diff
@@ -22,42 +22,74 @@ public class WikipediaComposeExtension
...
+            MessagingExtensionResponse composeExtensionResponse = null;
+            ImageResult imageResult = null;
+            List<MessagingExtensionAttachment> composeExtensionAttachments = new List<MessagingExtensionAttachment>();
+            var userData = await TemplateUtility.GetBotUserDataObject(_userState, turnContext, query);
+            var userPreferredCardType = userData.ComposeExtensionCardType;
+            var activity = turnContext.Activity;
             bool isSettingUrl = false;
 
-            var composeExtensionQuery = activity.GetComposeExtensionQueryData();
+            //var composeExtensionQuery = activity.GetComposeExtensionQueryData();
```
Comment (v-chetsh) [inline samples/app-complete-sample/csharp/template-bot-master-csharp/utility/WikipediaComposeExtension.cs:74]: // var composeExtensionQuery = activity.GetComposeExtensionQueryData();

File: samples/app-complete-sample/csharp/template-bot-master-csharp/utility/WikipediaComposeExtension.cs [hunk 15]
```diff
@@ -22,42 +22,74 @@ public class WikipediaComposeExtension
...
+            var queryParameter = GetQueryParameterByName(query, Strings.manifestParameterName);
 
             if (userData == null)
             {
-                composeExtensionResponse = new ComposeExtensionResponse();
-                string message = Strings.ComposeExtensionNoUserData;
-                composeExtensionResponse.ComposeExtension = GetMessageResponseResult(message);
-                return composeExtensionResponse;
+                composeExtensionResponse = new MessagingExtensionResponse();
+               string message = Strings.ComposeExtensionNoUserData;
```
Comment (v-chetsh) [inline samples/app-complete-sample/csharp/template-bot-master-csharp/utility/WikipediaComposeExtension.cs:90]: Alignment is not correct

**Review by v-chetsh — CHANGES_REQUESTED** [review]
Added my comments, please resolve.

**Review by v-chetsh — APPROVED** [review]
Changes looks good, approving it.

File: samples/app-complete-sample/csharp/README.md [hunk 16]
```diff
@@ -9,9 +9,10 @@ extensions:
...
   - Tabs
-  - Microsoft Bot Framework
+  - Microsoft Bot Framework v4
   createdDate: 9/22/2017 5:54:09 PM
-description: "Sample that shows how to build a bot for Microsoft Teams in C#."
+  updateDate:  
+description: "Sample that shows how to build a bot for Microsoft Teams in C# with bot framework v4."
 ---
 
 # Microsoft Teams Bot in C#
```
Comment (Wajeed-msft) [inline samples/app-complete-sample/csharp/README.md:18]: Could you please update the Git clone command? currently it's pointing to old repo. 

`git clone https://github.com/OfficeDev/microsoft-teams-template-bot-CSharp.git`

File: samples/app-complete-sample/csharp/README.md [hunk 17]
```diff
@@ -9,9 +9,10 @@ extensions:
   contentType: samples
   technologies:
   - Tabs
-  - Microsoft Bot Framework
+  - Microsoft Bot Framework v4
   createdDate: 9/22/2017 5:54:09 PM
-description: "Sample that shows how to build a bot for Microsoft Teams in C#."
+  updateDate:  
+description: "Sample that shows how to build a bot for Microsoft Teams in C# with bot framework v4."
```
Comment (Wajeed-msft) [inline samples/app-complete-sample/csharp/README.md:15]: Could you please include few screenshots?

**Review by Wajeed-msft — CHANGES_REQUESTED** [review]

File: samples/app-complete-sample/csharp/README.md (hunk 16 above)
Comment (NikitaJain-MSFT) [reply samples/app-complete-sample/csharp/README.md:18]: Updated git clone command

File: samples/app-complete-sample/csharp/README.md (hunk 17 above)
Comment (NikitaJain-MSFT) [reply samples/app-complete-sample/csharp/README.md:15]: Added screenshots of app features

**Review by Wajeed-msft — APPROVED** [review]

**Pull request merged**